### PR TITLE
[Algorithm] Support Rotated FCOS (ICCV'2019)

### PR DESCRIPTION
**Pull request by liuyanyi**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
#119 
#114 
https://github.com/jbwang1997/OBBDetection/blob/master/mmdet/models/dense_heads/fcos_head.py

## Modification
- [x]  Add RotatedAnchorFreeHead 
- [x]  Add RotatedFCOSHead
- [x]  Add CSLRFCOSHead
- [x]  Add RotatedIoULoss, need https://github.com/open-mmlab/mmcv/pull/1854
- [x] Add Doc
- [x]  Benchmark
- [x]  Publish model

Note: The result of Rotated IoU Loss is not stable, its still under tuning

## BC-breaking (Optional)
Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: configs/rotated_fcos/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+first time, we demonstrate a much simpler and flexible detection framework achieving improved detection accuracy. We
+hope that the proposed FCOS framework can serve as a simple and strong alternative for many other instance-level tasks.
+
+## Results and Models
+
+DOTA1.0
+
+|         Backbone         |  mAP  | Angle | Separate Angle | Tricks | lr schd | Mem (GB) | Inf Time (fps) | Aug | Batch Size |                                                Configs                                                |          Download          |
+|:------------------------:|:-----:|:-----:|:--------------:|:------:|:-------:|:--------:|:--------------:|:---:|:----------:|:-----------------------------------------------------------------------------------------------------:|:--------------------------:|
+| ResNet50 (1024,1024,200) | 70.70 | le90  |       Y        |   Y    |   1x    |    -     |       -        |  -  |     2      |    [rotated_fcos_sep_angle_r50_fpn_1x_dota_le90](./rotated_fcos_sep_angle_r50_fpn_1x_dota_le90.py)    | [model](-) &#124; [log](-) |
```
Comment (yangxue0827) [inline configs/rotated_fcos/README.md:29]: `Mem (GB)` is recorded in log.json file.
Inf Time (fps) can be test by:
`python -m torch.distributed.launch --nproc_per_node=1 --master_port=29500 tools/analysis_tools/benchmark.py YOUR_CFG YOUR_MODEL --launcher pytorch`

**Comment by zytx121** [comment]
Waiting for  `diff_iou_rotated_2d` in MMCV

File: mmrotate/models/losses/rotated_iou_loss.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import warnings
+
+import torch
+import torch.nn as nn
+from mmcv.ops import diff_iou_rotated_2d
```
Comment (zytx121) [inline mmrotate/models/losses/rotated_iou_loss.py:6]:
```
try:
    from mmcv.ops import diff_iou_rotated_2d
except:
    diff_iou_rotated_2d = None
```

File: mmrotate/models/losses/rotated_iou_loss.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        torch.Tensor: Loss tensor.
+    """
+    assert mode in ['linear', 'square', 'log']
+    if linear:
+        mode = 'linear'
+        warnings.warn(
+            'DeprecationWarning: Setting "linear=True" in '
+            'poly_iou_loss is deprecated, please use "mode=`linear`" '
+            'instead.')
+
```
Comment (zytx121) [inline mmrotate/models/losses/rotated_iou_loss.py:39]:
```
if diff_iou_rotated_2d is None:
        raise ImportError('Please install mmcv-full >= 1.5.0.')
```

File: configs/rotated_fcos/rotated_fcos_csl_gaussian_r50_fpn_1x_dota_le90.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            omega=1,
+            window='gaussian',
+            radius=1),
+        loss_cls=dict(
+            type='FocalLoss',
+            use_sigmoid=True,
+            gamma=2.0,
+            alpha=0.25,
+            loss_weight=1.0),
+        loss_bbox=dict(type='GIoULoss', loss_weight=1.0),
```
[the remark this answers is not in the document]
Comment (yangxue0827) [reply configs/rotated_fcos/rotated_fcos_csl_gaussian_r50_fpn_1x_dota_le90.py:26]: I guess it is horizontal GIoULoss

[the remark this answers is not in the document]
Comment (liuyanyi) [reply configs/rotated_fcos/rotated_fcos_csl_gaussian_r50_fpn_1x_dota_le90.py:26]: That's hbbox loss from mmdet

**Comment by yangxue0827** [comment]
This branch has conflicts that must be resolved.

Conflicting files
docs/en/model_zoo.md
docs/zh_cn/model_zoo.md

See details below.

**Review by yangxue0827 — APPROVED** [review]
LGTM
Waiting for models.

**Comment by zytx121** [comment]
Need to add metafile.yml, please refer to
https://github.com/open-mmlab/mmrotate/blob/main/configs/rotated_atss/metafile.yml
